### PR TITLE
Clean up test helpers.

### PR DESCRIPTION
**Pull request by davidmorgan**
Remove unused test helpers, inline test helpers only used once or that don't add enough.

### CODE REVIEW

File: _test_common/lib/builders.dart [hunk 1]
```diff
@@ -48,3 +36,40 @@ class DelegatingBuilder implements Builder {
   @override
   Future build(BuildStep buildStep) async => delegate.build(buildStep);
 }
+
+class PlaceholderBuilder extends Builder {
+  final String inputExtension;
+  final Map<String, String> outputExtensionsToContent;
```
Comment (lrhn) [inline _test_common/lib/builders.dart]: The "plural-to-singular" name confuses me.
(Also, what is "Content"? More docs could make this readable to someone who doesn't know the rest of the code. :wink: )

File: _test_common/lib/builders.dart [hunk 2]
```diff
@@ -48,3 +36,40 @@ class DelegatingBuilder implements Builder {
...
+
+  @override
+  Map<String, List<String>> get buildExtensions => {
+    inputExtension: outputExtensionsToContent.keys.toList(),
+  };
+
+  PlaceholderBuilder(
+    this.outputExtensionsToContent, {
```
Comment (lrhn) [inline _test_common/lib/builders.dart]: It takes a map as argument. Is it OK if that map changes afterwards? (If not, document that it shouldn't. If so, ... document that anyway.)

File: _test_common/lib/builders.dart [hunk 3]
```diff
@@ -48,3 +36,40 @@ class DelegatingBuilder implements Builder {
...
+    this.inputExtension = r'$lib$',
+  });
+
+  @override
+  Future build(BuildStep buildStep) async {
+    outputExtensionsToContent.forEach((extension, content) {
+      buildStep.writeAsString(
```
Comment (lrhn) [inline _test_common/lib/builders.dart]: Function is async, doesn't use `await`, calls an async `writeAsString` in a synchronous loop and doesn't await the result of that. Waiting for the returned future does not wait for the writing, which is *probably* intended for this API.

Either:
```dart
  for (final MapEntry(key: extension, value: content)) {
    await buildStep.writeAsString(
      _outputId(buildStep.inputId, inputExtension, extensions)),
      content,
    );
  }
```
or 
```dart
  await [for (final MapEntry(key: extension, value: content))
    buildStep.writeAsString(
      _outputId(buildStep.inputId, inputExtension, extensions)),
      content,
    )
  ].wait;
```

File: _test_common/lib/builders.dart [hunk 4]
```diff
@@ -48,3 +36,40 @@ class DelegatingBuilder implements Builder {
   @override
   Future build(BuildStep buildStep) async => delegate.build(buildStep);
 }
+
+class PlaceholderBuilder extends Builder {
+  final String inputExtension;
```
Comment (lrhn) [inline _test_common/lib/builders.dart]: The extension always includes the leading `.`, right?

File: _test_common/lib/common.dart [hunk 5]
```diff
@@ -25,40 +23,3 @@ Digest computeDigest(AssetId id, String contents) {
...
-    inputExtension: outputExtensionsToContent.keys.toList(),
-  };
-
-  PlaceholderBuilder(
-    this.outputExtensionsToContent, {
-    this.inputExtension = r'$lib$',
-  });
-
-  @override
-  Future build(BuildStep buildStep) async {
```
Comment (lrhn) [inline _test_common/lib/common.dart:44]: Ah, so you only moved the code, you dind't write it. (So technically OK to keep as-is). It's still suspicious hat it doen't await anything.

File: build_runner_core/test/environment/create_merged_dir_test.dart [hunk 6]
```diff
@@ -613,3 +612,71 @@ void _expectAllFiles(Directory dir) {
...
+class TestBuildEnvironment implements BuildEnvironment {
+  final TestReaderWriter _readerWriter;
+
+  @override
+  TestReaderWriter get reader => _readerWriter;
+
+  @override
+  TestReaderWriter get writer => _readerWriter;
+
+  /// If true, this will throw a [NonInteractiveBuildException] for all calls to
```
Comment (lrhn) [inline build_runner_core/test/environment/create_merged_dir_test.dart]:
```dart
  /// Whether to throw ...`
```

(Otherwise you have technically not said what `false` means :stuck_out_tongue:).

File: build_runner_core/test/environment/create_merged_dir_test.dart [hunk 7]
```diff
@@ -613,3 +612,71 @@ void _expectAllFiles(Directory dir) {
...
+  /// [prompt].
+  final bool throwOnPrompt;
+
+  /// The next response for calls to [prompt]. Must be set before calling
```
Comment (lrhn) [inline build_runner_core/test/environment/create_merged_dir_test.dart]: New paragraph after first sentence.

File: build_runner_core/test/environment/create_merged_dir_test.dart [hunk 8]
```diff
@@ -613,3 +612,71 @@ void _expectAllFiles(Directory dir) {
...
+  }
+
+  int? _nextPromptResponse;
+
+  TestBuildEnvironment({
+    TestReaderWriter? readerWriter,
+    this.throwOnPrompt = false,
+  }) : _readerWriter = readerWriter ?? TestReaderWriter();
+
+  /// Prompt the user for input.
```
Comment (lrhn) [inline build_runner_core/test/environment/create_merged_dir_test.dart]: "Promts ...."

File: build_runner_core/test/environment/create_merged_dir_test.dart [hunk 9]
```diff
@@ -613,3 +612,71 @@ void _expectAllFiles(Directory dir) {
...
+  /// chosen option is returned.
+  ///
+  /// If this environmment is non-interactive (such as when running in a test)
+  /// this method should throw [NonInteractiveBuildException].
+  @override
+  Future<int> prompt(String message, List<String> choices) {
+    if (throwOnPrompt) throw NonInteractiveBuildException();
+
+    assert(_nextPromptResponse != null);
+    return Future.value(_nextPromptResponse);
```
Comment (lrhn) [inline build_runner_core/test/environment/create_merged_dir_test.dart]: Consider writing `_nextPromptResponse!`.

The `Future.value` technically accepts `int?` as argument, but will throw if the value is `null`. Which it probably isn't if asserts are enabled.
I really, really want to disallow passing `null` as an argument when `T` is not nullable, so start preparing for this line to become a type error.

File: _test_common/lib/builders.dart (hunk 4 above)
Comment (davidmorgan) [reply _test_common/lib/builders.dart]: Please see next comment :)

File: _test_common/lib/builders.dart (hunk 1 above)
Comment (davidmorgan) [reply _test_common/lib/builders.dart]: This whole class is confusing: builder APIs talk in terms of file extensions, but "placeholders" are a special case, and this builder is just about placeholders.

So: renamed the variables to say what they actually do, then added a comment in the place where they are called "extensions" to explain that they're a special case of that API.

Thanks :)

File: _test_common/lib/builders.dart (hunk 2 above)
Comment (davidmorgan) [reply _test_common/lib/builders.dart]: Made it a BuiltMap, I like those ;)

File: _test_common/lib/builders.dart (hunk 3 above)
Comment (davidmorgan) [reply _test_common/lib/builders.dart]: Oh, I was thinking of getting rid of the `forEach`, a lint would have noticed that if I had.

I like the first one--done--thanks :)

File: _test_common/lib/common.dart (hunk 5 above)
Comment (davidmorgan) [reply _test_common/lib/common.dart:44]: Yup, I guess it just happened to work. It's actually only used in one test ... I guess it's not super well tested ;)

Anyway it's much clearer now, thanks. Placeholders are a bit of a corner case that people may not be aware of, so it's nice to explain clearly where we test it.

File: build_runner_core/test/environment/create_merged_dir_test.dart (hunk 6 above)
Comment (davidmorgan) [reply build_runner_core/test/environment/create_merged_dir_test.dart]: Yup yup, done. Thanks.

File: build_runner_core/test/environment/create_merged_dir_test.dart (hunk 7 above)
Comment (davidmorgan) [reply build_runner_core/test/environment/create_merged_dir_test.dart]: Done.

File: build_runner_core/test/environment/create_merged_dir_test.dart (hunk 8 above)
Comment (davidmorgan) [reply build_runner_core/test/environment/create_merged_dir_test.dart]: Done.

File: build_runner_core/test/environment/create_merged_dir_test.dart (hunk 9 above)
Comment (davidmorgan) [reply build_runner_core/test/environment/create_merged_dir_test.dart]: Done.

**Review by davidmorgan — COMMENTED** [review]
Thanks :)
